src/sql/src/query_model/rewrites/mod.rs

```diff
@@ -0,0 +1,165 @@
+}
+
+/// Classifies the rule by which boxes it should be applied to and in what
+/// order.
+#[allow(dead_code)]
+#[derive(Debug)]
+pub(crate) enum RuleType {
+    /// The rule should be applied to all boxes.
+    /// Order of the boxes does not matter.
+    AllBoxes,
+    /// The rule should be applied to all boxes in pre-order.
+    PreOrder,
+    /// The rule should be applied to all boxes in post-order.
+    PostOrder,
+    /// The rule should only be applied to the top box.
+    TopBoxOnly,
+}
+
+/// Generates blank instances of a rule.
+#[allow(missing_debug_implementations)]
+struct RuleGenerator {
+    /// The type of the rule.
+    pub typ: RuleType,
+    /// Function that creates a blank instance of the rule.
+    pub generate: Box<dyn Fn() -> Box<dyn Rule>>,
+}
+
+/// Create a RuleGenerator that generates blank instances of `R`.
+#[allow(dead_code)]
+fn make_generator<R: 'static + Rule + Default>() -> RuleGenerator {
+    RuleGenerator {
+        typ: R::rule_type(),
+        generate: Box::new(|| Box::new(R::default())),
+    }
+}
+
+/// Entry-point of the normalization stage.
+pub fn rewrite_model(model: &mut Model) {
+    // TODO: populate this list.
+    let mut rules: Vec<RuleGenerator> = vec![];
+    apply_rules_to_model(model, &mut rules);
+    model.garbage_collect();
+
+    // At the end of the process, update box and quantifier ids to make it
+    // easier to compare the graph before and after optimization.
+    model.update_ids();
+}
+
+/// Transform the model by applying a list of rewrite rules.
+fn apply_rules_to_model(model: &mut Model, rules: &mut Vec<RuleGenerator>) {
+    let mut rewritten = true;
+
+    while rewritten {
+        rewritten = false;
+
+        for rule in rules
+            .iter_mut()
+            .filter(|r| matches!(r.typ, RuleType::TopBoxOnly))
+        {
+            rewritten |= apply_rule(&rule, model, model.top_box);
+        }
+
+        rewritten |= recursively_apply_rules(rules, model, model.top_box, &mut HashSet::new());
+
+        for rule in rules
+            .iter_mut()
+            .filter(|r| matches!(r.typ, RuleType::TopBoxOnly))
+        {
+            rewritten |= apply_rule(&rule, model, model.top_box);
+        }
+    }
+}
+
+/// Applies the rewrite rule corresponding to the `rule_generator` to a box.
+///
+/// Returns whether the box was rewritten.
+fn apply_rule(rule_generator: &RuleGenerator, model: &mut Model, box_id: BoxId) -> bool {
+    // Create a fresh instance of the rule.
+    let mut rule = (*rule_generator.generate)();
+    let rewrite = rule.check(model, box_id);
+    if rewrite {
+        rule.rewrite(model, box_id);
+    }
+    rewrite
+}
+
+/// Apply the given list of rewrite rules to the given box and
+/// all descendent boxes.
+///
+/// `visited_boxes` keeps track of all the visited boxes so far
+/// to avoid visiting them again.
+///
+/// Returns whether any box was rewritten.
+fn recursively_apply_rules(
```

Comment:
This needs to be stack-safe. We have stack other stack-safety issues in QGM and we need to discuss how to address them.